api/git/azure.go

```diff
@@ -294,3 +396,189 @@ func getVersionType(name string) string {
 	}
 	return "commit"
 }
+
+func (a *azureClient) listRefs(ctx context.Context, opt option) ([]string, error) {
+	config, err := parseUrl(opt.repositoryUrl)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to parse url")
+	}
+
+	listRefsUrl, err := a.buildRefsUrl(config)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to build list refs url")
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "GET", listRefsUrl, nil)
+	if opt.username != "" || opt.password != "" {
+		req.SetBasicAuth(opt.username, opt.password)
+	} else if config.username != "" || config.password != "" {
+		req.SetBasicAuth(config.username, config.password)
+	}
+
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to create a new HTTP request")
+	}
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to make an HTTP request")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		if resp.StatusCode == http.StatusNotFound {
+			return nil, ErrIncorrectRepositoryURL
+		} else if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusNonAuthoritativeInfo {
+			return nil, ErrAuthenticationFailure
+		}
+		return nil, fmt.Errorf("failed to list refs url with a status \"%v\"", resp.Status)
+	}
+
+	var refs struct {
+		Value []azureRef
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&refs); err != nil {
+		return nil, errors.Wrap(err, "could not parse Azure refs response")
+	}
+
+	var ret []string
+	for _, value := range refs.Value {
+		if value.Name == "HEAD" {
+			continue
+		}
+		ret = append(ret, value.Name)
+	}
+
+	if a.cacheEnabled {
+		a.repoRefCache.Add(opt.repositoryUrl, ret)
+	}
+	return ret, nil
+}
+
+func (a *azureClient) listFiles(ctx context.Context, opt option) ([]string, error) {
+	var (
+		allPaths    []string
+		filteredRet []string
+		refs        []string
+		err         error
+		success     bool
+		ok          bool
+		cache       interface{}
+	)
```

Comment:
There is no need to allocate all the vars that could be used, not only it would be better from a memory allocation point, but also would help address some "funny" conditional logic like the one on line 493, where `ok` and `success` may not be even initilised before that.

Reply:
By default the value of the boolean variable is `false`. The logic here is that if either `ok` or `success` is not `true`, we will call `listRefs`. In this case, only when the key pair is found in the cache before line 493, `ok` and `success` can be updated to `true`. 
There were not so many allocated vars, but I added them according to the change of cache type (pointer)